### ISSUE TITLE
GradientBoostingRegressor performance issue

### ISSUE THREAD

**Issue by jiaxincao** [catboost/catboost#1415]
Problem: GradientBoostingRegressor is much slower than sklearn.ensemble.GradientBoostingRegressor
catboost version: 0.24.1
Operating System: Ubuntu 20.04.1
CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz

I'm trying to use CatBoostRegressor to train a model. However, I found the following two problems:
* CatBoostRegressor is converging to a reasonably small loss value (comparing to sklearn.ensemble.GradientBoostingRegressor)
* CatBoostRegressor is running very slow (comparing to sklearn.ensemble.GradientBoostingRegressor)

## Code to repro

Here are the code to repro the issue.

On the CatBoostRegressor side:

```
from catboost import Pool, CatBoostClassifier, CatBoostRegressor
import math
import random

random.seed(0)

train_data = [1000 * random.random() for _ in range(1000)]
train_label = [1000 * random.random() for _ in range(1000)]

train_dataset = Pool(
    data=train_data,
    label=train_label,
)

model = CatBoostRegressor(
    iterations=1000,
    max_depth=16,
    learning_rate=0.1,
    loss_function="MAE",
    metric_period=100,
)
model.fit(train_dataset)

error = 0
for x, y in zip(train_data, train_label):
    error += math.fabs(y - model.predict([x]))
print(error)
```

On the sklearn side:
```
from sklearn.ensemble import GradientBoostingRegressor
import numpy as np
import random
import math

random.seed(0)

train_data = [[1000 * random.random()] for _ in range(1000)]
train_label = [1000 * random.random() for _ in range(1000)]

model = GradientBoostingRegressor(
    iterations=1000,
    max_depth=16,
    learning_rate=0.1,
    loss="lad",
    verbose=True,
)
model.fit(train_data, train_label)

error = 0
for x, y in zip(train_data, train_label):
    error += math.fabs(y - model.predict([x])[0])
print(error)
```

## Results Comparison

Running the CatBoostRegressor gives me:

```
0:      learn: 248.9398445      total: 72.4ms   remaining: 1m 12s
100:    learn: 221.8620448      total: 13s      remaining: 1m 55s
200:    learn: 205.8289189      total: 28.4s    remaining: 1m 52s
300:    learn: 200.6325808      total: 43.8s    remaining: 1m 41s
400:    learn: 198.5311027      total: 59.2s    remaining: 1m 28s
500:    learn: 197.4161546      total: 1m 14s   remaining: 1m 14s
600:    learn: 196.9112287      total: 1m 29s   remaining: 59.7s
700:    learn: 196.6157447      total: 1m 45s   remaining: 44.9s
800:    learn: 196.3736242      total: 2m       remaining: 30s
900:    learn: 196.2290317      total: 2m 16s   remaining: 14.9s
999:    learn: 196.1664007      total: 2m 31s   remaining: 0us
196166.40133150056

real    2m38.181s
user    2m39.796s
sys     0m3.246s
```

Running sklearn.ensemble.GradientBoostingRegressor gives me:

```
Iter       Train Loss   Remaining Time
         1         237.7821           27.43s
         2         228.3763           26.79s
         3         222.3447           23.71s
         4         216.0993           23.00s
         5         213.7406           20.26s
         6         211.9449           18.38s
         7         210.5194           16.99s
         8         209.5082           15.64s
         9         208.6418           14.60s
        10         207.9725           13.72s
        20         184.2427           12.21s
        30         163.0125           12.84s
        40         161.0923           10.75s
        50         160.7532            9.31s
        60         160.0838            8.38s
        70         154.9149            8.25s
        80         149.4753            8.12s
        90         135.7374            8.05s
       100         126.5245            8.01s
       200          98.5639            7.26s
       300          79.9776            6.91s
       400          64.7833            6.42s
       500          57.6789            5.73s
       600          45.0907            4.81s
       700          39.9052            3.76s
       800          35.2271            2.57s
       900          32.9738            1.31s
      1000          31.4612            0.00s
31461.20087697405

real    0m14.620s
user    0m14.722s
sys     0m0.373s
```

Comparing the two results, I found:
* sklearn took about 14.6s to train the model, while CatBoostRegressor took 2m28s
* sklearn's loss evaluation is 31461.20, while CatBoostRegressor is 196166.40

Apparently, the results suggested that sklearn was training faster, and yield to better performance.

Could anyone help take a look? Is there anything that I missed when using CatBoostRegressor? Is there anything that I could do to make it train faster and also get better model performance?

**Comment by kizill** [comment]
At first, in your example there is no actual information to "learn", as there is no correlation between label and features. So, the model is actually trying to memorize values for individual features. Second, CatBoost uses feature values quantization on 254 buckets by default, and in each bucket label values are distribute uniformely, thus for scoring function two buckets are practically the same. So, to memorize each object you have to disable any generalization abilities of algorithm and to overfit on such a random data :)
## How to overfit CatBoost on train?
So, to overfit CatBoost on this data as fast as you can, try this settings:
```python
model = CatBoostRegressor(
# for 1000 object 100 trees is enough
        iterations=100,
# go as deep as you can
        max_depth=16,
# by default, CatBoost uses SymmetricTree grow policy, that itself brings a form of regularization to training
        grow_policy='Depthwise',
# we are just memorizing leaf values, so learning rate near to 1 is good enough
        learning_rate=0.9999999,
# here we are allowing CatBoost to make quantized feature-bins for each individual value in dataset
        border_count=1000,
        loss_function="MAE",
# disable adding random noise in split scores
        random_strength=0,
# use all objects on each iteration! no bootrap - no problems
        bootstrap_type='No',
# and of course disable regularization :)
        l2_leaf_reg=0.0,
        min_data_in_leaf=1
    )
```

**Comment by jiaxincao** [comment]
Thanks, @kizill !
